iOS/Source/LoginController.swift

```diff
@@ -168,36 +168,21 @@ class LoginController: SweetTableController {
     func didPressSignInButton() {
         let username = self.loginParameters.username ?? ""
         let password = self.loginParameters.password ?? ""
-        let base64Credentials = "\(username):\(password)".data(using: .utf8)!.base64EncodedString()
-        let authField = "Basic \(base64Credentials)"
-
-        let url = URL(string: "https://api.pinboard.in/v1/user/api_token?format=json")!
-        let request = URLRequest(url: url)
-
-        let config = URLSessionConfiguration.default
-        config.httpAdditionalHeaders = ["Authorization": authField]
-        self.session = URLSession(configuration: config)
-
-        self.session.dataTask(with: request) { (data, response, error) in
-            DispatchQueue.main.async {
-                guard let data = data else { return }
-                if (response as? HTTPURLResponse)?.statusCode == 401 {
-                    let alert = UIAlertController.dismissableAlert(title: "Error", message: "Username/password combination invalid!")
-                    self.present(alert, animated: true)
-                }
 
-                do {
-                    guard let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: String] else { return }
-                    if let result = json["result"] {
-                        self.fetcher.userToken = "\(username):\(result)"
-                        self.dismiss(animated: true)
-                    }
-                }
-                catch {
-                    
+        self.fetcher.networking.authenticate(username: username, password: password)
+        self.fetcher.login() { (JSON, error) in
+            if let error = error {
+                var message = error.localizedDescription
+                if error.code == 401 {
+                    message = "Username/password combination invalid!"
                 }
+                let alert = UIAlertController.dismissableAlert(title: "Authentation failed", message: message)
+                self.present(alert, animated: true)
+            } else if let result = JSON?["result"] {
+                self.fetcher.userToken = "\(username):\(result)"
```

Comment:
This should be done inside fetcher. The Controller shouldn't have knowledge about this.